### PR TITLE
Adding 2nd version of DTK::initialize()

### LINKED ISSUE #254: Provide the second version of DTK_initialize
Based on `Kokkos::InitArguments` in [Kokkos_Core.hpp](https://github.com/kokkos/kokkos/blob/master/core/src/Kokkos_Core.hpp). Also need it for C_API (which would probably need to copy the parameters). Must have for Fortran.

### PR DESCRIPTION
**Pull request by aprokop**
Fix #254.

Need to add some tests.

### CODE REVIEW

File: packages/Utils/src/DTK_Core.cpp [hunk 1]
```diff
@@ -25,8 +23,8 @@ bool dtkIsInitialized = false;
 bool dtkInitializedKokkos = false;
 
 // Initialize Kokkos, if it needs initialization.
-// This takes the same arguments as (the first two of) initialize().
-void initKokkos( int *argc, char ***argv )
+template <typename... Args>
+void initKokkos( Args... args )
```
Comment (dalg24) [inline packages/Utils/src/DTK_Core.cpp]: So no `&&` here, ...

File: packages/Utils/src/DTK_Core.cpp [hunk 2]
```diff
@@ -39,7 +37,7 @@ void initKokkos( int *argc, char ***argv )
         if ( !kokkosIsInitialized )
         {
             // Unlike MPI_Init, Kokkos promises not to modify argc and argv.
-            Kokkos::initialize( *argc, *argv );
+            Kokkos::initialize( std::forward<Args &>( args )... );
```
Comment (dalg24) [inline packages/Utils/src/DTK_Core.cpp]: ..., and you added a `&` here.

Would you please comment that we could not use perfect forwarding and explain what is happening there?
Or did you try each combination until the compiler stopped complaining?

File: packages/Utils/src/DTK_Core.cpp [hunk 3]
```diff
@@ -52,12 +50,14 @@ void initKokkos( int *argc, char ***argv )
...
 
-void initialize( int *argc, char ***argv )
+template <typename... Args>
+void initialize( Args &&... args )
 {
     if ( !dtkIsInitialized )
-        initKokkos( argc, argv );
+        initKokkos( std::forward<Args>( args )... );
     dtkIsInitialized = true;
 }
```
Comment (dalg24) [inline packages/Utils/src/DTK_Core.cpp]: This one is textbook perfect forwarding, right?

File: packages/Interface/src/DTK_C_API.h [hunk 4]
```diff
@@ -53,7 +53,10 @@ extern bool DTK_is_valid( DTK_UserApplicationHandle handle );
 extern void DTK_destroy( DTK_UserApplicationHandle handle );
 
 /*! Initialize DTK */
-extern void DTK_initialize( int *argc, char ***argv );
+extern void DTK_initialize();
+// FIXME: provide a version based on full Kokkos::InitArguments
+// FIXME: can we do something with the name here?
+extern void DTK_initialize_cmd( int *argc, char ***argv );
```
Comment (dalg24) [inline packages/Interface/src/DTK_C_API.h]: Why did you add the suffix?

Comment (aprokop) [inline packages/Interface/src/DTK_C_API.h]: No overload in C...

File: packages/Utils/src/DTK_Core.cpp (hunk 3 above)
Comment (aprokop) [inline packages/Utils/src/DTK_Core.cpp]: Yes. But I had to change the instantiation and the actual call.

File: packages/Utils/src/DTK_Core.cpp (hunk 1 above)
Comment (aprokop) [inline packages/Utils/src/DTK_Core.cpp]: > Or did you try each combination until the compiler stopped complaining?

That! It may not be the only solution...

File: packages/Utils/src/DTK_Core.cpp (hunk 2 above)
Comment (aprokop) [inline packages/Utils/src/DTK_Core.cpp]: > Or did you try each combination until the compiler stopped complaining?

That! It may not be the only solution...

File: packages/Interface/src/DTK_C_API.h (hunk 4 above)
Comment (dalg24) [inline packages/Interface/src/DTK_C_API.h]: Oh that's right...

File: packages/Utils/src/DTK_Core.cpp [hunk 5]
```diff
@@ -74,4 +74,10 @@ void finalize()
 
     dtkIsInitialized = false;
 }
+
+// ETI for initialize
+template void initialize<int &, char **&>( int &argc, char **&argv );
```
Comment (dalg24) [inline packages/Utils/src/DTK_Core.cpp]: Can you comment on that one?
I just looked it up and found:
`void Kokkos::initialize(int& narg, char* arg[]);`

**Comment by aprokop** [comment]
Hmm, question about writing tests for this: if we are not allowed to `initialize` and `finalize` Kokkos multiple times in the same run, do I need to write 2+ files to test this functionality?

**Comment by aprokop** [comment]
I should actually just merge `tstInit1-3` into a single test with a command line switch so that it does pollute the directory.

**Comment by aprokop** [comment]
Ready for final review.

File: packages/Interface/test/tstInit.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+ * the LICENSE file in the top-level directory.                             *
+ ****************************************************************************/
+/*!
+ * \file
+ * \brief  Kokkos initialization tests
+ */
+
+#include "DTK_Core.hpp"
+
+#include <getopt.h>
```
Comment (dalg24) [inline packages/Interface/test/tstInit.cpp]: Should we be using `#include <unistd.h>`?

File: packages/Interface/test/tstInit.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    if ( t == 1 )
+    {
+        status = ( status &&
+                   !( DataTransferKit::isInitialized() ||
+                      Kokkos::DefaultExecutionSpace::is_initialized() ) );
+
+        DataTransferKit::initialize( &argc, &argv );
+        status = ( status &&
+                   !( !DataTransferKit::isInitialized() ||
+                      !Kokkos::DefaultExecutionSpace::is_initialized() ) );
```
Comment (dalg24) [inline packages/Interface/test/tstInit.cpp]: Seriously??
```
status == ( status && ( DataTransferKit::isInitialized() && Kokkos::DefaultExecutionSpace::is_initialized() ) );
```

File: packages/Interface/test/tstInit.cpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                   !( DataTransferKit::isInitialized() ||
+                      Kokkos::DefaultExecutionSpace::is_initialized() ) );
+    }
+    else if ( t == 2 )
+    {
+        Kokkos::initialize( argc, argv );
+
+        status = ( status &&
+                   !( DataTransferKit::isInitialized() ||
+                      !Kokkos::DefaultExecutionSpace::is_initialized() ) );
```
Comment (dalg24) [inline packages/Interface/test/tstInit.cpp]:
```
... && ( !DataTransferKit::isInitialized() && Kokkos::DefaultExecutionSpace::is_initialized() )
```

File: packages/Interface/test/tstInit.cpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+        DataTransferKit::initialize( &argc, &argv );
+        status = ( status &&
+                   !( !DataTransferKit::isInitialized() ||
+                      !Kokkos::DefaultExecutionSpace::is_initialized() ) );
```
Comment (dalg24) [inline packages/Interface/test/tstInit.cpp]: Change

File: packages/Interface/test/tstInit.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+        DataTransferKit::finalize();
+        status = ( status &&
+                   !( DataTransferKit::isInitialized() ||
+                      !Kokkos::DefaultExecutionSpace::is_initialized() ) );
```
Comment (dalg24) [inline packages/Interface/test/tstInit.cpp]: Change

File: packages/Interface/test/tstInit.cpp (hunk 6 above)
Comment (aprokop) [inline packages/Interface/test/tstInit.cpp]: Maybe... But this one works too.

File: packages/Interface/test/tstInit.cpp (hunk 7 above)
Comment (aprokop) [inline packages/Interface/test/tstInit.cpp]: Due to Coverage... Wanted to merge `if()` statement into a single line with `status`.

**Comment by aprokop** [comment]
After a brief chat with @dalg24 , decided to change the `status` checks to something like this:
```c++
auto check = lambda[status&](bool cond) {
    status = ( status && !cond );

check(DataTransferKit::isInitialized() || Kokkos::DefaultExecutionSpace::is_initialized());
```

**Comment by aprokop** [comment]
OK, addressed the comments.
